Cart/milestone 1/3

Evo-Eclipse opened the pull request:
### Overview

This pull request implements milestone 1-3 for the shopping cart feature in iOS-FakeNFT app. The changes introduce network layer infrastructure, data models, services for cart management, currency selection and payment processing, along with core UI components and supporting assets.

### Changes

Network Models & Services

- Added `Currency`, `Order`, `OrderUpdate`, and `PaymentResult` models for API communication
- Implemented `CurrencyService`, `OrderService`, and `PaymentService` with corresponding request builders
- Updated `ServicesAssembly` to register new services

UI Components

- Added `CartCardView` for displaying shopping cart items
- Added `CurrencyCardView` for currency selection interface
- Added `DeleteConfirmationView` for cart item deletion confirmation

Assets

- Added cryptocurrency icons (Bitcoin, Ethereum, Dogecoin, Cardano, Solana, Tether, ApeCoin, ShibaInu)
- Added 4 NFT placeholder images for mockup purposes

### Documentation

- Added `Cart.md` with shopping cart implementation plan

### Motivation

This milestone establishes the foundational architecture for shopping cart functionality, enabling users to manage NFT purchases, select payment currencies, and process orders through a clean, modular service layer with reusable UI components.

---

### Extras

- Task tracker link — [FakeNFT - SwiftUI - MVVM](https://github.com/users/suhanovv/projects/2/views/4)
- Epic decomposition document — [Cart.md](https://github.com/suhanovv/iOS-FakeNFT-StarterProject-Extended-Public/blob/18d8dead90dc4444f8496bc98e12e737d7b2e415/docs/Cart.md)
- Architecture — MVVM
- UI Framework — SwiftUI

Evo-Eclipse committed:
docs: add shopping cart implementation plan

Evo-Eclipse committed:
feat: add cryptocurrency icons to asset catalog

- Create Currency Icons asset catalog
- Add ApeCoin (APE) icon in @2x and @3x resolutions
- Add Bitcoin (BTC) icon in @2x and @3x resolutions
- Add Cardano (ADA) icon in @2x and @3x resolutions
- Add Dogecoin (DOGE) icon in @2x and @3x resolutions
- Add Ethereum (ETH) icon in @2x and @3x resolutions
- Add Shiba Inu (SHIB) icon in @2x and @3x resolutions
- Add Solana (SOL) icon in @2x and @3x resolutions
- Add Tether (USDT) icon in @2x and @3x resolutions

Evo-Eclipse committed:
feat: add NFT placeholder images to asset catalog

- Create Mockup Images asset catalog with namespace
- Add NFTPlaceholder1 image asset
- Add NFTPlaceholder2 image asset
- Add NFTPlaceholder3 image asset
- Add NFTPlaceholder4 image asset

Evo-Eclipse committed:
feat: add shopping cart network layer and services

- Add Currency, Order, PaymentResult, and OrderUpdate models
- Add CurrencyRequests for fetching currencies
- Add OrderRequests for getting and updating orders
- Add PaymentRequests for processing payments
- Add CurrencyService with get currencies and get currency by id methods
- Add OrderService with get and update order methods
- Add PaymentService with payment processing method
- Register new services in ServicesAssembly

Evo-Eclipse committed:
feat: add CartCardView component for shopping cart items

- Add CartCardView with image, name, rating, price, and delete button
- Add StarView subview with 5-star rating display
- Add preview with sample NFT data
- Use 108x108 image with 12pt corner radius
- Add delete button with cart delete icon
- Support rating normalization between 0-5

Evo-Eclipse committed:
feat: add CurrencyCardView component for currency selection

- Add CurrencyCardView with image, title, code, and selection state
- Add 36x36 image with 6pt corner radius
- Add title in black and code in green text
- Add selection border with 1pt stroke
- Add default and selected state previews with Bitcoin example

Evo-Eclipse committed:
feat: add DeleteConfirmationView component for cart item deletion

- Add DeleteConfirmationView with image, confirmation text, and action buttons
- Add 108x108 image with 12pt corner radius
- Add confirmation message in footnote font
- Add delete button with red text and cancel button with white text
- Add preview with NFT placeholder image

Evo-Eclipse commented:
### Branching note

#### Current Situation

The `cart/epic-branch` was created from `develop` but does not contain any commits of its own before `cart/milestone-1-3` branched off. Both `cart/epic-branch` and `develop` currently point to the same commit `1fa8525`.

#### Analysis of Other Epics

I've reviewed the branching structure across all epics in this project:

| Epic Branch | Has Own Commits Before Milestones | Status |
|-------------|-----------------------------------|--------|
| `profile/epic` | Yes (`c97d05a` - "Added profile decomposition to docs") | Matches reference |
| `catalogue/epic-branch` | No (only merge commits from milestones) | Same as cart |
| `statistics/epic-branch` | No (only merge commits from milestones) | Same as cart |
| `cart/epic-branch` | No (milestone work started directly) | **Current PR** |

#### Reference Git flow

```mermaid
%%{init: { 'theme': 'base', 'gitGraph': {'showCommitLabel':false}} }%%
gitGraph
    commit tag: "main"
    commit
    branch develop
    checkout develop
    commit
    commit
    branch epic_name
    checkout epic_name
    commit tag: "empty commit?"
    commit tag: "empty commit?"
    branch sprint_19_20
    checkout sprint_19_20
    commit
    commit
    checkout epic_name
    merge sprint_19_20 tag: "merge sprint → epic"
    commit
    checkout develop
    merge epic_name tag: "merge epic → develop"
    checkout main
    merge develop
```

#### Expected Git flow

```mermaid
%%{init: { 'theme': 'base', 'gitGraph': {'showCommitLabel':false}} }%%
gitGraph
    commit tag: "main"
    commit
    branch develop
    checkout develop
    commit
    commit
    branch cart/epic-branch
    checkout cart/epic-branch
    branch cart/milestone-1-3
    checkout cart/milestone-1-3
    commit
    commit tag: "HEAD (current state)"
    checkout cart/epic-branch
    merge cart/milestone-1-3 tag: "future merge milestone → epic"
    commit
    checkout develop
    merge cart/epic-branch tag: "future merge epic → develop"
    checkout main
    merge develop
```

#### Current Git status
```mermaid
%%{init: { 'theme': 'base', 'gitGraph': {'showCommitLabel':false}} }%%
gitGraph
    commit tag: "main"
    commit
    branch develop
    checkout develop
    commit
    commit tag: "cart/epic-branch"
    branch cart/milestone-1-3
    checkout cart/milestone-1-3
    commit
    commit
    commit tag: "HEAD (current state)"
```

#### Question

Should I add an initial commit to `cart/epic-branch` (e.g., update Cart.md documentation) to align with the reference architecture, or is the current structure acceptable given that 2 out of 3 other epics follow the same pattern?

The branching hierarchy itself is correct: `main` → `develop` → `cart/epic-branch` → `cart/milestone-1-3`.

Evo-Eclipse committed:
docs: update milestone 1 completion times for shopping cart

- Update milestone 1 fact time from pending to 8h 15m
- Add completion times for all milestone 1 tasks
- Total time under estimate: 8h 15m vs 19h estimated

IgorChumikov commented on iOS-FakeNFT-Extended/Models/Network/Currency.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+struct Currency: Decodable {
```
Так как Currency является DTO-моделью (используется для передачи данных и парсинга ответа),
важно, чтобы любой разработчик мог быстро понять назначение каждого поля,
не заглядывая в API или бэкенд-документацию.

👉 Рекомендуется добавить документацию (///) к каждому свойству.

Это:

улучшает читаемость
упрощает онбординг коллег
```

import Foundation

/// DTO-модель валюты, получаемая из сетевого ответа
struct Currency: Decodable {
    /// Уникальный идентификатор валюты
    let id: String
    /// Отображаемое название валюты (например: "Доллар США")
    let title: String
    /// Короткое или системное имя валюты (например: "USD")
    let name: String
    /// URL изображения (иконка / флаг валюты)
    let image: URL
}

```

IgorChumikov commented on iOS-FakeNFT-Extended/Services/CurrencyService.swift:
```diff
@@ -0,0 +1,33 @@
+
+import Foundation
+
+protocol CurrencyService {
```
По названию выглядит как конкретная реализация (класс/сервис),
хотя по факту это абстракция (protocol).

Это может вводить в заблуждение коллег при чтении кода.

🔹 Рекомендуемый вариант (Swift-style)
```

protocol CurrencyServiceProtocol {
    func getCurrencies() async throws -> [Currency]
    func getCurrency(id: String) async throws -> Currency
}

```
🔹  префикс I — это стиль C# / Java

```

protocol ICurrencyService {
    func getCurrencies() async throws -> [Currency]
    func getCurrency(id: String) async throws -> Currency
}

```

IgorChumikov commented on iOS-FakeNFT-Extended/Services/CurrencyService.swift:
```diff
@@ -0,0 +1,33 @@
+}
+
+@MainActor
+final class CurrencyServiceImpl: CurrencyService {
```
Так как CurrencyServiceImpl — это сервис, а не UI-компонент,
его лучше объявить как actor, а не помечать @MainActor.

@MainActor означает:

весь код сервиса всегда выполняется на главном потоке
каждый сетевой запрос блокирует main thread
это UI-контекст, а не сервисный

👉 @MainActor подходит для:

ViewModel
ViewController
обновления UI

❌ Но не для сетевых / бизнес-сервисов

✅ Почему actor — правильное решение

actor:

это тот же класс, но
потокобезопасный по умолчанию
защищает состояние от data race
отлично подходит для сервисов, которые:
делают сеть
кешируют данные
используются из разных потоков

👉 По сути:
actor = class + thread safety
```

actor CurrencyServiceImpl: CurrencyServiceProtocol {

    private let networkClient: NetworkClient

    init(networkClient: NetworkClient) {
        self.networkClient = networkClient
    }

    func getCurrencies() async throws -> [Currency] {
        let request = GetCurrenciesRequest()
        return try await networkClient.send(request: request)
    }

    func getCurrency(id: String) async throws -> Currency {
        let request = GetCurrencyByIdRequest(id: id)
        return try await networkClient.send(request: request)
    }
}

```

IgorChumikov commented on iOS-FakeNFT-Extended/Services/PaymentService.swift:
```diff
@@ -0,0 +1,29 @@
+}
+
+@MainActor
+final class PaymentServiceImpl: PaymentService {
```
actor

IgorChumikov commented on iOS-FakeNFT-Extended/Services/PaymentService.swift:
```diff
@@ -0,0 +1,29 @@
+
+import Foundation
+
+protocol PaymentService {
```
PaymentServiceProtocol

IgorChumikov commented on iOS-FakeNFT-Extended/UI/Cart/Views/CartCardView.swift:
```diff
@@ -0,0 +1,95 @@
+    let priceText: String
+    let onDelete: () -> Void
+
+    // MARK: - View
```
// MARK: - Body

IgorChumikov commented on iOS-FakeNFT-Extended/UI/Cart/Views/CartCardView.swift:
```diff
@@ -0,0 +1,95 @@
+
+import SwiftUI
+
+struct CartCardView: View {
```
В проекте есть отдельная фича “Корзина (Cart)”, поэтому код, относящийся к ней,
лучше структурировать по feature-подходу, а не хранить вперемешку с другими экранами.

✅ Рекомендуется
Создать отдельную папку Cart, внутри которой уже разложить код по слоям:
```

Cart
├── Views
├── ViewModels
└── Models

```

Почему так лучше:

сразу видно, что относится к корзине
проще ориентироваться новым разработчикам
масштабируемо при росте фичи

IgorChumikov commented on iOS-FakeNFT-Extended/UI/Components/CurrencyCardView.swift:
```diff
@@ -0,0 +1,68 @@
+
+import SwiftUI
+
+struct CurrencyCardView: View {
```
❌ CurrencyCardView

Название слишком общее — не очевидно, что карточка используется именно в корзине.

✅ Более удачные варианты:

CartCurrencyCardView ← самый понятный
CartPaymentCurrencyView
CartCurrencySelectionView

👉 Рекомендуемый вариант:
```
CartCurrencyCardView
```

IgorChumikov commented on iOS-FakeNFT-Extended/UI/Components/DeleteConfirmationView.swift:
```diff
@@ -0,0 +1,65 @@
+
+import SwiftUI
+
+struct DeleteConfirmationView: View {
```
❌ DeleteConfirmationView

Название универсальное, но контекст — удаление из корзины,
это важно отразить в имени.

✅ Более удачные варианты:

CartDeleteConfirmationView ← рекомендуемый
CartItemDeleteConfirmationView
CartRemoveItemConfirmationView

👉 Рекомендуемый вариант:
```
CartDeleteConfirmationView
```

IgorChumikov commented on iOS-FakeNFT-Extended/UI/Cart/Views/CartDeleteConfirmationView.swift:
```diff
@@ -0,0 +1,65 @@
+    let image: Image
+    let onDelete: () -> Void
+    let onCancel: () -> Void
+
```
Итог

Так как в проекте есть фича Cart, рекомендуется:

вынести весь связанный код в папку Cart
внутри разделить на Views / ViewModels / Models
использовать Cart в названиях UI-компонентов
(например, CartCurrencyCardView, CartDeleteConfirmationView)

Это улучшит читаемость, навигацию и поддержку кода.

Evo-Eclipse committed:
docs: add documentation comments to Currency model

- Add struct-level documentation describing it as network DTO model
- Add documentation for model fields

Evo-Eclipse committed:
refactor: convert CurrencyService and PaymentService to actors

- Rename CurrencyService protocol to CurrencyServiceProtocol
- Rename CurrencyServiceImpl to CurrencyService
- Rename PaymentService protocol to PaymentServiceProtocol
- Rename PaymentServiceImpl to PaymentService
- Replace @MainActor with actor for thread-safe concurrent access
- Update ServicesAssembly to use new protocol names

Evo-Eclipse committed:
refactor: reorganize cart components into dedicated module structure

- Move CartCardView from Components to Cart/Views
- Rename CurrencyCardView to CartCurrencyCardView and move to Cart/Views
- Rename DeleteConfirmationView to CartDeleteConfirmationView and move to Cart/Views
- Create Cart module with Models, ViewModels, and Views folders
- Update Xcode project file references and build phases

Evo-Eclipse committed:
feat: add dynamic button width synchronization to CartDeleteConfirmationView

- Add buttonWidth state to track maximum button width
- Add ButtonWidthPreferenceKey to propagate width changes
- Replace fixed 127pt width with dynamic width based on content
- Update horizontal padding from 16pt to 22pt
- Use GeometryReader to measure button content width
- Synchronize both buttons to use maximum width for consistent sizing
